File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function (s, k) {
    let left = 0, maxCount = 0;
    const freq = new Map();
    let res = 0;

    for (let right = 0; right < s.length; right++) {
        const char = s[right];
        freq.set(char, (freq.get(char) || 0) + 1);
        maxCount = Math.max(maxCount, freq.get(char));

        if (right - left + 1 - maxCount > k) {
            freq.set(s[left], freq.get(s[left]) - 1);
            left++;
        }
        res = Math.max(res, right - left + 1);
    }

    return res;

};

// claude chat link
// https://claude.ai/chat/50d71e5d-216f-4028-bc01-738567d1283e

// claude animation link
// https://claude.ai/artifacts/70bc2c17-825f-4bf4-a0ce-8e084ccc6a1d

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        left = 0
        max_count = 0
        freq = {}
        res = 0
        
        for right in range(len(s)):
            char = s[right]
            freq[char] = freq.get(char, 0) + 1
            max_count = max(max_count, freq[char])
            
            if right - left + 1 - max_count > k:
                freq[s[left]] = freq[s[left]] - 1
                left += 1
            
            res = max(res, right - left + 1)
        
        return res